From 32154c0466bcc525bbffd5dd87a73ec21128e8b7 Mon Sep 17 00:00:00 2001
From: Xavier Hanin <xavier@apache.org>
Date: Fri, 30 Sep 2005 15:54:19 +0000
Subject: [PATCH] FIX: ivy is now able to use simple ivy files in cache
 (doesn't need resolver info, use default one if no resolver is given)
 (IVY-86)

git-svn-id: https://svn.apache.org/repos/asf/incubator/ivy/trunk@484044 13f79535-47bb-0310-9956-ffa450edef68
---
 CHANGES.txt                                |  1 +
 src/java/fr/jayasoft/ivy/Ivy.java          |  4 ++++
 test/java/fr/jayasoft/ivy/ResolveTest.java | 20 ++++++++++++++++++++
 test/java/fr/jayasoft/ivy/ivy-mod1.2.xml   |  8 ++++++++
 4 files changed, 33 insertions(+)
 create mode 100644 test/java/fr/jayasoft/ivy/ivy-mod1.2.xml

diff --git a/CHANGES.txt b/CHANGES.txt
index 07d49cf47..5934bb8ad 100644
--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,3 +1,4 @@
+- FIX: ivy is now able to use simple ivy files in cache (doesn't need resolver info, use default one if no resolver is given) (IVY-86)
 - FIX: private conf not accessible from other modules (IVY-76)
 - FIX: root module configurations isolation bug fixed (IVY-84)
 - FIX: changed the place where ivy stores master ivy files in cache to not overlap
diff --git a/src/java/fr/jayasoft/ivy/Ivy.java b/src/java/fr/jayasoft/ivy/Ivy.java
index 594fbb942..31ea6eacc 100644
--- a/src/java/fr/jayasoft/ivy/Ivy.java
+++ b/src/java/fr/jayasoft/ivy/Ivy.java
@@ -966,6 +966,10 @@ public ResolvedModuleRevision findModuleInCache(ModuleRevisionId mrid, File cach
                 try {
                     ModuleDescriptor depMD = XmlModuleDescriptorParser.parseDescriptor(this, ivyFile.toURL(), validate);
                     DependencyResolver resolver = (DependencyResolver)_resolversMap.get(depMD.getResolverName());
+                    if (resolver == null) {
+                        Message.debug("\tresolver not found: "+depMD.getResolverName()+" => trying to use default one for "+mrid);                                    
+                        resolver = getDefaultResolver();
+                    }
                     if (resolver != null) {
                         Message.debug("\tfound ivy file in cache for "+mrid+": "+ivyFile);
                         return new DefaultModuleRevision(resolver, depMD, false, false);
diff --git a/test/java/fr/jayasoft/ivy/ResolveTest.java b/test/java/fr/jayasoft/ivy/ResolveTest.java
index 1fb677685..33bcd6178 100644
--- a/test/java/fr/jayasoft/ivy/ResolveTest.java
+++ b/test/java/fr/jayasoft/ivy/ResolveTest.java
@@ -75,6 +75,26 @@ public void testResolveSimple() throws Exception {
         assertTrue(_ivy.getArchiveFileInCache(_cache, "org1", "mod1.2", "2.0", "mod1.2", "jar", "jar").exists());
     }
 
+    public void testFromCacheOnly() throws Exception {
+        Ivy ivy = new Ivy();
+        ivy.configure(new File("test/repositories/bugIVY-56/ivyconf.xml"));
+        
+//        ResolveReport report = ivy.resolve(new File("test/repositories/1/org1/mod1.1/ivys/ivy-1.0.xml").toURL(),
+//                null, new String[] {"*"}, _cache, null, true);
+//        // should have an error, the conf is bad and the dependency should not be found
+//        assertTrue(report.hasError());
+
+        // put necessary stuff in cache, and it should now be ok
+        File ivyfile = ivy.getIvyFileInCache(_cache, ModuleRevisionId.newInstance("org1", "mod1.2", "2.0"));
+        File art = ivy.getArchiveFileInCache(_cache, "org1", "mod1.2", "2.0", "mod1.2", "jar", "jar");
+        FileUtil.copy(ResolveTest.class.getResource("ivy-mod1.2.xml"), ivyfile, null);
+        FileUtil.copy(new File("test/repositories/1/org1/mod1.2/jars/mod1.2-2.0.jar"), art, null);
+
+        ResolveReport report = ivy.resolve(new File("test/repositories/1/org1/mod1.1/ivys/ivy-1.0.xml").toURL(),
+                null, new String[] {"*"}, _cache, null, true);
+        assertFalse(report.hasError());
+    }
+
     public void testChangeCacheLayout() throws Exception {
         Ivy ivy = new Ivy();
         ivy.configure(new File("test/repositories/ivyconf.xml"));
diff --git a/test/java/fr/jayasoft/ivy/ivy-mod1.2.xml b/test/java/fr/jayasoft/ivy/ivy-mod1.2.xml
new file mode 100644
index 000000000..d0dffd67c
--- /dev/null
+++ b/test/java/fr/jayasoft/ivy/ivy-mod1.2.xml
@@ -0,0 +1,8 @@
+<ivy-module version="1.0">
+	<info organisation="org1"
+	       module="mod1.2"
+	       revision="2.0"
+	       status="integration"
+	       publication="20041101110000"
+	/>
+</ivy-module>
